atecc608a_se.c

```diff
@@ -0,0 +1,224 @@
+    ASSERT_STATUS(atcab_init(&atca_iface_config), ATCA_SUCCESS);
+    ASSERT_STATUS(atcab_is_locked(LOCK_ZONE_CONFIG, &config_locked), ATCA_SUCCESS);
+    printf("  - Config locked: %d\n", config_locked);
+    ASSERT_STATUS(atcab_is_locked(LOCK_ZONE_DATA, &locked), ATCA_SUCCESS);
+    printf("  - Data locked: %d\n", locked);
+    for(uint8_t i=0; i < 16; i++)
+    {
+        ASSERT_STATUS(atcab_is_slot_locked(i, &locked), ATCA_SUCCESS);
+        printf("  - Slot %d locked: %d\n", i, locked);
+    }
+    ASSERT_STATUS(atcab_release(), ATCA_SUCCESS);
+    printf("--------------------------------\n");
+    if(!config_locked)
+    {
+        printf("Config is not locked! Cannot use private keys.\n");
+        return ATCA_NOT_LOCKED;
+    }
+    return ATCA_CONFIG_ZONE_LOCKED;
+}
+
+psa_status_t atecc608a_export_key(psa_key_slot_number_t key,
+                                         uint8_t *p_data,
+                                         size_t data_size,
+                                         size_t *p_data_length)
+{
+    ATCA_STATUS ret = ATCA_SUCCESS;
+    uint16_t slot = key;
+    size_t key_data_len = 65;
+    if(data_size < key_data_len)
+    {
+        printf("FAILED! Key buffer too small.\n");
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+    ASSERT_STATUS(atcab_init(&atca_iface_config), ATCA_SUCCESS);
+
+    ret = atcab_get_pubkey(slot, &p_data[1]);
+    if (ret != ATCA_SUCCESS)
+    {
+        printf("FAILED! atcab_get_pubkey: %d\n", ret);
+        return atecc608a_to_psa_error(ret);
+    }
+    p_data[0] = 4;
+    *p_data_length = key_data_len;
+    printf("atecc608a_export_key - pubkey size %d:\n", *p_data_length);
+    atcab_printbin_sp(p_data, *p_data_length);
+    ASSERT_STATUS(atcab_release(), ATCA_SUCCESS);
+}
+
+psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
+                                       psa_algorithm_t alg,
+                                       const uint8_t *p_hash,
+                                       size_t hash_length,
+                                       uint8_t *p_signature,
+                                       size_t signature_size,
+                                       size_t *p_signature_length)
+{
+    ATCA_STATUS ret = ATCA_SUCCESS;
+    uint16_t key_id = key_slot;
+
+    /* We can only do ECDSA on SHA-256 */
+    /* PSA_ALG_ECDSA(PSA_ALG_SHA_256) */
+    if(!PSA_ALG_IS_ECDSA(alg))
+    {
+        printf("FAILED! PSA_ALG_IS_ECDSA:\n");
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (hash_length != 32)
+    {
+        printf("FAILED! hash_length:\n");
+        /* The driver only supports signing things of length 32. */
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (!PSA_ALG_IS_RANDOMIZED_ECDSA(alg))
+    {
+        printf("FAILED! PSA_ALG_IS_RANDOMIZED_ECDSA:\n");
+        /* The hardware only supports randomized ECDSA */
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    /* XXX how to check the curve? */
```

Comment:
What if the key is always pre-provisioned and never created by any API call?

We have a function to register a slot as occupied (different from creating the key), so should this fail if we try to say a key slot is occupied by a key the hardware doesn't support? Should the driver be notified when a new key is specified as being pre-provisioned at run-time so that it can error if it doesn't support that type of key?

Reply:
The pre-provisioning mechanism needs to set the metadata that the crypto library uses.

A key can't be “specified as being pre-provisioned at run-time”. By definition, a pre-provisioned key is one that exists before the system boots. If a key is provisioned at runtime, this should be done via the normal key creation mechanism.

Mbed Crypto doesn't need to be aware of what the driver is capable to handle pre-provisioned keys: that's the problem of the pre-provisioning mechanism. Mbed Crypto does need to know what slots are capable of storing what kinds of keys (type, size). I don't think we've fully nailed down the mechanism for that yet. This driver hard-codes slot assignments, so it doesn't need this slot capability mechanism.

Reply:
A key can be “specified as being pre-provisioned at run-time”. You say "Hey Mbed Crypto. I have a key pre-provisioned in this slot, associated with this lifetime. You don't have to generate it or import it. It's already in that slot and the hardware knows about it" via `psa_register_se_slot()`. This was the proposed API from Derek developed during prototyping. The API may not include such a function yet, but this example assumes that it exists; we may want to change this.

Reply:
Oh, I see what you mean. The pre-provisioning happened earlier but the specification of it happens at run time. Yes, indeed. But in any case, that mechanism must set all the metadata about the key, including its type. The library core stores the metadata outside the SE.

Reply:
Yes, and I think that function is the right place to reject the addition of invalid metadata, like an unsupported key type. Which means we need a driver hook for being able to provide the rejection criteria, giving the driver the opportunity to reject metadata.

Reply:
But no metadata is added in this function! This function doesn't take a key type as argument. When an application performs a signature with a pre-existing key, it never needs to know the key type. This function takes an algorithm as argument, and the core guarantees that this algorithm is compatible with the slot's policy. The place where the key type and the policy can and must be checked against what the slot contains is when the key is registered.

Reply:
`psa_register_se_slot()` does take the key type as an argument.